Update Migrate Web api topic plus samples to 3.1 (#18417)
dotnet/AspNetCore.Docs: aspnetcore/migration/webapi.md

description: Learn how to migrate a web API implementation from ASP.NET 4.x Web API to ASP.NET Core MVC.
ms.author: scaddie
ms.custom: mvc
ms.date: 05/20/2020
no-loc: [Blazor, "Identity", "Let's Encrypt", Razor, SignalR]
uid: migration/webapi
---
# Migrate from ASP.NET Web API to ASP.NET Core

By [Scott Addie](https://twitter.com/scott_addie) and [Steve Smith](https://ardalis.com/)

An ASP.NET 4.x Web API is an HTTP service that reaches a broad range of clients, including browsers and mobile devices. ASP.NET Core unifies ASP.NET 4.x's MVC and Web API app models into a unified programming model known as ASP.NET Core MVC. This article demonstrates the steps required to migrate from ASP.NET 4.x Web API to ASP.NET Core MVC.

[View or download sample code](https://github.com/dotnet/AspNetCore.Docs/tree/master/aspnetcore/migration/webapi/sample) ([how to download](xref:index#how-to-download-a-sample))

> ::: moniker range=">= aspnetcore-3.0"

Reviewer: The current version ***ALWAYS*** goes first:

* In page links are broken in the 2nd and following moniker sections.
* It's the first thing customers (and us to) see when editing the file.

Author: @Rick-Anderson, thanks. The latest version is first in this case.  The content breakdown of versioning is 3.0 or higher on the top and version 2.2 or lower located at the bottom. The content at the top in this case is for  3.0 and 3.1.   There wasn't a 3.1 specific instruction or information, beyond the 3.1 download link provided in the include.  Thanks for pointing out the bad links, embarrassed I missed that.

## Prerequisites

[!INCLUDE [prerequisites](../includes/net-core-prereqs-vs-3.1.md)]

## Review ASP.NET 4.x Web API project

> As a starting point, this article uses the *ProductsApp* project created in [Getting Started with ASP.NET Web API 2](/aspnet/web-api/overview/getting-started-with-aspnet-web-api/tutorial-your-first-web-api). In that project, a basic ASP.NET 4.x Web API project is configured as follows.

Reviewer: ~As a starting point,~ this article uses the *ProductsApp* project created in [Getting Started with ASP.NET Web API 2](/aspnet/web-api/overview/getting-started-with-aspnet-web-api/tutorial-your-first-web-api). In that project, a basic ASP.NET 4.x Web API project is configured as follows.

Less is more

In *Global.asax.cs*, a call is made to `WebApiConfig.Register`:

[!code-csharp[](webapi/sample/3.x/ProductsApp/Global.asax.cs?highlight=14)]

The `WebApiConfig` class is found in the *App_Start* folder and has a static `Register` method:

[!code-csharp[](webapi/sample/3.x/ProductsApp/App_Start/WebApiConfig.cs)]

This class configures [attribute routing](/aspnet/web-api/overview/web-api-routing-and-actions/attribute-routing-in-web-api-2), although it's not actually being used in the project. It also configures the routing table, which is used by ASP.NET 4.x Web API. In this case, ASP.NET 4.x Web API expects URLs to match the format `/api/{controller}/{id}`, with `{id}` being optional.

> The *ProductsApp* project includes one controller. The controller inherits from `ApiController` and contains two actions:

Reviewer: Too verbose. Why tell them it includes 1 controller? Don't repeat info in the other tutorial. Folks doing a migration are well versed in ASP.NET 4.x

All you need is the following:
The `ProductsController`:

No need to tell them  inherits from ApiController, they know that from where they're coming from.

[!code-csharp[](webapi/sample/3.x/ProductsApp/Controllers/ProductsController.cs?highlight=28,33)]

> The `Product` model used by `ProductsController` is a basic class:

Reviewer: Why tell them it's a basic class?

The `ProductsController` returns `Product` model objects:

Reviewer: On 2nd thought, don't even say that. This tutorial should only be about the conversion. You'll need some info like how routing worked the old way and how it works now. Assume the audience is well versed in Web API .NET 4.x and just needs help on the migration.

The following sections demonstrate migration of the Web API project to ASP.NET Core MVC.

## Create the destination project

Create a new blank solution in Visual Studio and add the ASP.NET 4.x Web API project to migrate:

1. From the **File** menu, select **New** > **Project**.
1. Select the **Blank Solution** template and select **Next**.
1. Name the solution *WebAPIMigration*. Select **Create**.
1. Add the existing *ProductsApp* project to the solution.

Add a new API project to migrate to:

1. Add a new **ASP.NET Core Web Application** project to the solution.
1. In the **Configure your new project** dialog, Name the project *ProductsCore*, and select **Create**.
1. In the **Create a new ASP.NET Core Web Application** dialog, confirm that **.NET Core** and **ASP.NET Core 3.1** are selected. Select the **API** project template, and select **Create**.
1. Remove the *WeatherForecast.cs* and *Controllers/WeatherForecastController.cs* example files from the new *ProductsCore* project.

The solution now contains two projects. The following sections explain migrating the *ProductsApp* project's contents to the *ProductsCore* project.

## Migrate configuration

> ASP.NET Core doesn't use the *App_Start* folder or the *Global.asax* file, and the *web.config* file is added at publish time. *Startup.cs* is the replacement for *Global.asax* and is located in the project root. The `Startup` class handles all app startup tasks. For more information, see <xref:fundamentals/startup>.

Reviewer: Way too long.

ASP.NET Core doesn't use:

* *App_Start* folder or the *Global.asax* file.
* *web.config* is typically not used and is only added when publishing.

The `Startup` class:

* Replaces *Global.asax* 
*  handles all app startup tasks

No need to tell them it's in the root, they'll figure that out.

> All ASP.NET Core project templates include attribute routing in the generated code. The following `UseRouting` and `UseEndpoints` calls register route matching and endpoint execution in the [middleware](xref:fundamentals/middleware/index) pipeline. They replace the *ProductsApp* project's *App_Start/WebApiConfig.cs* file:

Reviewer: The following `UseRouting` and `UseEndpoints` calls:

* register route matching and endpoint execution in the [middleware](xref:fundamentals/middleware/index) pipeline. {comment make route matching  a link to routing}
* Replace *App_Start/WebApiConfig.cs* file

> All ASP.NET Core project templates include attribute routing in the generated code. The following `UseRouting` and `UseEndpoints` calls register route matching and endpoint execution in the [middleware](xref:fundamentals/middleware/index) pipeline. They replace the *ProductsApp* project's *App_Start/WebApiConfig.cs* file:

Reviewer 2: I haven't had time to fix one thing with that tool ... the asterisk and backtick should be URL encoded. I'll try to get that tool updated this weekend.

> All ASP.NET Core project templates include attribute routing in the generated code. The following `UseRouting` and `UseEndpoints` calls register route matching and endpoint execution in the [middleware](xref:fundamentals/middleware/index) pipeline. They replace the *ProductsApp* project's *App_Start/WebApiConfig.cs* file:

Reviewer 2: I updated the app from 2.2 to 3.1, and it now furnishes the links with encoded asterisks and backticks. **_BONUS_**: Thanks to @scottaddie's tip, it now provides a *name with type* link. :tada:

I'll be using it over the weekend to do the Blazor API doc cross-links, so I'll see if it runs ok. Let me know if you hit a 🐞 with it. :ear:

<img width="616" alt="Capture" src="https://user-images.githubusercontent.com/1622880/82732474-08ecf180-9cd3-11ea-8d6c-6de5f65f3723.PNG">

For more information, see <xref:fundamentals/startup>.

## Migrate models and controllers

> The following code shows the `ProductsController` update for ASP.NET Core:

Reviewer 3: ```suggestion
The following code shows the `ProductsController` to be updated for ASP.NET Core:
```

> Copy over the *ProductApp* project's controller and the model it uses. Follow these steps:

Reviewer: ~Copy over the *ProductApp* project's controller and the model it uses. Follow these steps:~

1. Copy *Controllers/ProductsController.cs* from the original project to the new one.

> 1. Copy the entire *Models* folder from the original project to the new one.

Reviewer: 1. Copy the ~entire~ *Models* folder from the original project to the new one.

> 1. Change the copied files' namespaces to match the new project name (*ProductsCore*). Adjust the `using ProductsApp.Models;` statement in *ProductsController.cs* too.

Reviewer: 1. Change copied files'  root namespace to `whatever the new root namespace is` 
1. Update the using statement from `orig using` to `new using`.
What's the trailing too all about?

> At this point, building the app results in a number of compilation errors. The errors occur because the following components don't exist in ASP.NET Core:

Reviewer: Why are you telling them this in the middle of the migration. Just keep going.

Author: It was original to the instruction that was already reviewed, as many of these were so I left it in.  I actually found it helpful since the first thing I did out of habit was compile to see what shape they were in and ran into the errors, then saw that was expected.  It is likely most folks will not do that however and assume they just need to keep going through steps and not be concerned.  So I made the changes.

1. Update the `using ProductsApp.Models;` statement to `using ProductsCore.Models;`.

The following components don't exist in ASP.NET Core:

* `ApiController` class
* `System.Web.Http` namespace
* `IHttpActionResult` interface

Make the following changes:

1. Change `ApiController` to <xref:Microsoft.AspNetCore.Mvc.ControllerBase>. Add `using Microsoft.AspNetCore.Mvc;` to resolve the `ControllerBase` reference.
1. Delete `using System.Web.Http;`.
1. Change the `GetProduct` action's return type from `IHttpActionResult` to `ActionResult<Product>`.

> Simplify the `GetProduct` action's `return` statement to the following:

Reviewer 3: This step should be part of the numbered list above it. Be sure to indent the associated code snippet once you move it there (because you want the snippet to be visually associated with the step).

```csharp
return product;
```

## Configure routing

> All ASP.NET Core project templates include attribute routing in the generated code.

Reviewer 3: I suggest future-proofing this statement by talking specifically about the API project template. The project templates change quite often.

```suggestion
The ASP.NET Core *API* project template includes endpoint routing configuration in the generated code.
```

The following sections demonstrate migration of the Web API project to ASP.NET Core MVC.

## Create the destination project

Create a new blank solution in Visual Studio and add the ASP.NET 4.x Web API project to migrate:

1. From the **File** menu, select **New** > **Project**.
1. Select the **Blank Solution** template and select **Next**.
1. Name the solution *WebAPIMigration*. Select **Create**.
1. Add the existing *ProductsApp* project to the solution.

Add a new API project to migrate to:

1. Add a new **ASP.NET Core Web Application** project to the solution.
1. In the **Configure your new project** dialog, Name the project *ProductsCore*, and select **Create**.
1. In the **Create a new ASP.NET Core Web Application** dialog, confirm that **.NET Core** and **ASP.NET Core 3.1** are selected. Select the **API** project template, and select **Create**.
1. Remove the *WeatherForecast.cs* and *Controllers/WeatherForecastController.cs* example files from the new *ProductsCore* project.

The solution now contains two projects. The following sections explain migrating the *ProductsApp* project's contents to the *ProductsCore* project.

## Migrate configuration

> ASP.NET Core doesn't use:
> 
> * *App_Start* folder or the *Global.asax* file
> * *web.config* file is added at publish time.

Reviewer 3: ```suggestion
ASP.NET Core doesn't use the *App_Start* folder or the *Global.asax* file. Additionally, the *web.config* file is added at publish time.
```

The `Startup` class:

* Replaces *Global.asax*.
* Handles all app startup tasks.

For more information, see <xref:fundamentals/startup>.

## Migrate models and controllers

The following code shows the `ProductsController` update for ASP.NET Core:

> [!code-csharp[](webapi/sample/3.x/ProductsApp/Controllers/ProductsController.cs)]

Reviewer 3: ```suggestion

[!code-csharp[](webapi/sample/3.x/ProductsApp/Controllers/ProductsController.cs)]
```

Update the `ProductsController` for ASP.NET Core:

> 1. Copy *Controllers/ProductsController.cs* from the original project to the new one.
> 1. Copy the *Models* folder from the original project to the new one.

Reviewer 3: Consider combining these two steps into one:
```suggestion
1. Copy *Controllers/ProductsController.cs* and the *Models* folder from the original project to the new one.
```

1. Update the `using ProductsApp.Models;` statement to `using ProductsCore.Models;`.

The following components don't exist in ASP.NET Core:

* `ApiController` class
* `System.Web.Http` namespace
* `IHttpActionResult` interface

Make the following changes:

1. Change `ApiController` to <xref:Microsoft.AspNetCore.Mvc.ControllerBase>. Add `using Microsoft.AspNetCore.Mvc;` to resolve the `ControllerBase` reference.
1. Delete `using System.Web.Http;`.
1. Change the `GetProduct` action's return type from `IHttpActionResult` to `ActionResult<Product>`.

Simplify the `GetProduct` action's `return` statement to the following:

```csharp
return product;
```

## Configure routing

All ASP.NET Core project templates include attribute routing in the generated code.

The following <xref:Microsoft.AspNetCore.Builder.EndpointRoutingApplicationBuilderExtensions.UseRouting%2A> and <xref:Microsoft.AspNetCore.Builder.EndpointRoutingApplicationBuilderExtensions.UseEndpoints%2A> calls:

* Register route matching and endpoint execution in the [middleware](xref:fundamentals/middleware/index) pipeline.
* Replace the *ProductsApp* project's *App_Start/WebApiConfig.cs* file.

> [!code-csharp[](webapi/sample/3.x/ProductsCore/Startup.cs?name=snippet_Configure&highlight=10,14)]

Reviewer 3: I recommend moving this code snippet and its association explanation to the numbered list below. It should be the first step in the list because it's required to use routing.

Author: @scottaddie, good point, thanks, but there really isn't a real step to take with UseRouting and UseEndPoints.  It is just already generated with the project.

Is this what you were envisioning?

## Configure routing

1. All ASP.NET Core project templates include attribute routing in the generated code.

The following <xref:Microsoft.AspNetCore.Builder.EndpointRoutingApplicationBuilderExtensions.UseRouting%2A> and <xref:Microsoft.AspNetCore.Builder.EndpointRoutingApplicationBuilderExtensions.UseEndpoints%2A> calls:

* Register route matching and endpoint execution in the [middleware](xref:fundamentals/middleware/index) pipeline.
* Replace the *ProductsApp* project's *App_Start/WebApiConfig.cs* file.

[!code-csharp[](webapi/sample/3.x/ProductsCore/Startup.cs?name=snippet_Configure&highlight=10,14)]

1. Mark the `ProductsController` class with the following attributes:

    ```csharp
    [Route("api/[controller]")]
    [ApiController]
    ```

Reviewer 3: @wadepickett Disregard my comment. I forgot that we tell the reader to create a new project.

description: Learn how to migrate a web API implementation from ASP.NET 4.x Web API to ASP.NET Core MVC.
ms.author: scaddie
ms.custom: mvc
ms.date: 05/20/2020
no-loc: [Blazor, "Identity", "Let's Encrypt", Razor, SignalR]
uid: migration/webapi
---
# Migrate from ASP.NET Web API to ASP.NET Core

By [Scott Addie](https://twitter.com/scott_addie) and [Steve Smith](https://ardalis.com/)

> An ASP.NET 4.x Web API is an HTTP service that reaches a broad range of clients, including browsers and mobile devices. ASP.NET Core unifies ASP.NET 4.x's MVC and Web API app models into a single programming model known as ASP.NET Core MVC. This article demonstrates the steps required to migrate from ASP.NET 4.x Web API to ASP.NET Core MVC.

Reviewer 3: ```suggestion
An ASP.NET 4.x Web API is an HTTP service that reaches a broad range of clients, including browsers and mobile devices. ASP.NET Core unifies ASP.NET 4.x's MVC and Web API app models into a unified programming model known as ASP.NET Core MVC. This article demonstrates the steps required to migrate from ASP.NET 4.x Web API to ASP.NET Core MVC.
```

[View or download sample code](https://github.com/dotnet/AspNetCore.Docs/tree/master/aspnetcore/migration/webapi/sample) ([how to download](xref:index#how-to-download-a-sample))

::: moniker range=">= aspnetcore-3.0"

## Prerequisites

[!INCLUDE [prerequisites](../includes/net-core-prereqs-vs-3.1.md)]

## Review ASP.NET 4.x Web API project

As a starting point, this article uses the *ProductsApp* project created in [Getting Started with ASP.NET Web API 2](/aspnet/web-api/overview/getting-started-with-aspnet-web-api/tutorial-your-first-web-api). In that project, a basic ASP.NET 4.x Web API project is configured as follows.

In *Global.asax.cs*, a call is made to `WebApiConfig.Register`:

[!code-csharp[](webapi/sample/3.x/ProductsApp/Global.asax.cs?highlight=14)]

The `WebApiConfig` class is found in the *App_Start* folder and has a static `Register` method:

[!code-csharp[](webapi/sample/3.x/ProductsApp/App_Start/WebApiConfig.cs)]

> This class configures [attribute routing](/aspnet/web-api/overview/web-api-routing-and-actions/attribute-routing-in-web-api-2), although it's not actually being used in the project. It also configures the routing table, which is used by ASP.NET Web API. In this case, ASP.NET 4.x Web API expects URLs to match the format `/api/{controller}/{id}`, with `{id}` being optional.

Reviewer 3: ```suggestion
This class configures [attribute routing](/aspnet/web-api/overview/web-api-routing-and-actions/attribute-routing-in-web-api-2), although it's not actually being used in the project. It also configures the routing table, which is used by ASP.NET 4.x Web API. In this case, ASP.NET 4.x Web API expects URLs to match the format `/api/{controller}/{id}`, with `{id}` being optional.
```

The *ProductsApp* project includes one controller. The controller inherits from `ApiController` and contains two actions:

[!code-csharp[](webapi/sample/3.x/ProductsApp/Controllers/ProductsController.cs?highlight=28,33)]

The `Product` model used by `ProductsController` is a basic class:

[!code-csharp[](webapi/sample/3.x/ProductsApp/Models/Product.cs)]

The following sections demonstrate migration of the Web API project to ASP.NET Core MVC.

## Create the destination project

> Create a new blank solution in Visual Studio and add the Web API project to migrate:

Reviewer 3: ```suggestion
Create a new blank solution in Visual Studio and add the ASP.NET 4.x Web API project to migrate:
```

1. From the **File** menu, select **New** > **Project**.

> 1. Select the **Blank Solution** template and click **Next**.

Reviewer 3: ```suggestion
1. Select the **Blank Solution** template and select **Next**.
```

> 1. Name the solution *WebAPIMigration*. Click **Create**.

Reviewer 3: Avoid the word "click", as it implies a mouse must be used.

```suggestion
1. Name the solution *WebAPIMigration*. Select **Create**.
```

1. Add the existing *ProductsApp* project to the solution.

Add a new API project to migrate to:

1. Add a new **ASP.NET Core Web Application** project to the solution.

> 1. In the **Configure your new project** dialog, Name the project *ProductsCore*, and click **Create**.

Reviewer 3: ```suggestion
1. In the **Configure your new project** dialog, Name the project *ProductsCore*, and select **Create**.
```

> 1. In the **Create a new ASP.NET Core Web Application** dialog, confirm that **.NET Core** and **ASP.NET Core 3.1** are selected. Select the **API** project template, and click **Create**.

Reviewer 3: ```suggestion
1. In the **Create a new ASP.NET Core Web Application** dialog, confirm that **.NET Core** and **ASP.NET Core 3.1** are selected. Select the **API** project template, and select **Create**.
```

1. Remove the *WeatherForecast.cs* and *Controllers/WeatherForecastController.cs* example files from the new *ProductsCore* project.

The solution now contains two projects. The following sections explain migrating the *ProductsApp* project's contents to the *ProductsCore* project.

## Migrate configuration

ASP.NET Core doesn't use the *App_Start* folder or the *Global.asax* file, and the *web.config* file is added at publish time. *Startup.cs* is the replacement for *Global.asax* and is located in the project root. The `Startup` class handles all app startup tasks. For more information, see <xref:fundamentals/startup>.

> All ASP.NET Core templates include attribute routing in the generated code. The following `UseRouting` and `UseEndpoints` calls register route matching and endpoint execution in the [middleware](xref:fundamentals/middleware/index) pipeline.  They replace the *ProductsApp* project's *App_Start/WebApiConfig.cs* file:

Reviewer 3: ```suggestion
All ASP.NET Core project templates include attribute routing in the generated code. The following `UseRouting` and `UseEndpoints` calls register route matching and endpoint execution in the [middleware](xref:fundamentals/middleware/index) pipeline. They replace the *ProductsApp* project's *App_Start/WebApiConfig.cs* file:
```

> [!code-csharp[](webapi/sample/3.x/ProductsCore/Startup.cs?name=snippet_Configure&highlight=10,14])]

Reviewer 3: The line highlighting doesn't appear to be working.

At this point, building the app results in a number of compilation errors. The errors occur because the following components don't exist in ASP.NET Core:

* `ApiController` class
* `System.Web.Http` namespace
* `IHttpActionResult` interface

Fix the errors as follows:

1. Change `ApiController` to <xref:Microsoft.AspNetCore.Mvc.ControllerBase>. Add `using Microsoft.AspNetCore.Mvc;` to resolve the `ControllerBase` reference.
1. Delete `using System.Web.Http;`.
1. Change the `GetProduct` action's return type from `IHttpActionResult` to `ActionResult<Product>`.

Simplify the `GetProduct` action's `return` statement to the following:

```csharp
return product;
```

## Configure routing

Configure routing as follows:

1. Mark the `ProductsController` class with the following attributes:

    ```csharp
    [Route("api/[controller]")]
    [ApiController]
    ```

    The preceding [`[Route]`](xref:Microsoft.AspNetCore.Mvc.RouteAttribute) attribute configures the controller's attribute routing pattern. The [`[ApiController]`](xref:Microsoft.AspNetCore.Mvc.ApiControllerAttribute) attribute makes attribute routing a requirement for all actions in this controller.

>     Attribute routing supports tokens, such as `[controller]` and `[action]`. At runtime, each token is replaced with the name of the controller or action, respectively, to which the attribute has been applied. The tokens reduce the number of magic strings in the project. The tokens also ensure routes remain synchronized with the corresponding controllers and actions when automatic rename refactorings are applied.

Reviewer 3: ```suggestion
    Attribute routing supports tokens, such as `[controller]` and `[action]`. At runtime, each token is replaced with the name of the controller or action, respectively, to which the attribute has been applied. The tokens:
    
    * Reduce the number of magic strings in the project.
    * Ensure routes remain synchronized with the corresponding controllers and actions when automatic rename refactorings are applied.
```

> 1. Set the project's compatibility mode to ASP.NET Core 3.0:

Reviewer 3: This step should be removed. It has no affect in 3.0+ apps. Be sure to remove the corresponding code from the sample app too.

    [!code-csharp[](webapi/sample/3.x/ProductsCore/Startup.cs?name=snippet_ConfigureServices&highlight=4)]

    The preceding change:

    * Is required to use the `[ApiController]` attribute at the controller level.
    * Opts in to potentially breaking behaviors introduced in ASP.NET Core 3.0.
1. Enable HTTP Get requests to the `ProductController` actions:
    * Apply the [`[HttpGet]`](xref:Microsoft.AspNetCore.Mvc.HttpGetAttribute) attribute to the `GetAllProducts` action.
    * Apply the `[HttpGet("{id}")]` attribute to the `GetProduct` action.

After the preceding changes and the removal of unused `using` statements, *ProductsController.cs* file looks like this:

[!code-csharp[](webapi/sample/3.x/ProductsCore/Controllers/ProductsController.cs)]

Run the migrated project, and browse to `/api/products`. A full list of three products appears. Browse to `/api/products/1`. The first product appears.

## Additional resources

* <xref:web-api/index>
* <xref:web-api/action-return-types>
* <xref:mvc/compatibility-version>

> ::: moniker-end

Reviewer 3: ```suggestion
::: moniker-end

```

> After the preceding changes and the removal of unused `using` statements, *ProductsController.cs* file looks like this:

Author: @Rick-Anderson, I moved the finished controller to show that first.  Good point.  I did not skip the steps to replace with only changes after. While it is cleaner, in the context the topic this section is preceded and followed by steps for the migration. I feel like in this case it makes it a bit harder to follow by dropping two steps in the middle.
